src/tree/mod.rs

```diff
@@ -0,0 +1,128 @@
+use std::io;
+
+use dusk_bls12_381::Scalar;
+use kelvin::{
+    annotation,
+    annotations::{Cardinality, Count},
+    Branch, BranchMut, ByteHash, Compound, Content,
+};
+use nstack::NStack;
+
+use crate::merkle_lvl_hash::hash::*;
+use crate::merkle_proof::poseidon_branch::extend_scalar;
+use crate::ARITY;
+use crate::{PoseidonBranch, PoseidonLevel, StorageScalar};
+
+annotation! {
+    /// The annotation for the PoseidonTree
+    pub struct PoseidonAnnotation {
+        scalar: StorageScalar,
+        count: Cardinality<u64>,
+    }
+}
+
+/// A zk-friendly datastructure to store elements
+pub struct PoseidonTree<T, H>
+where
+    T: Content<H>,
+    for<'any> &'any T: Into<StorageScalar>,
+    H: ByteHash,
+{
+    branch_depth: usize,
+    inner: NStack<T, PoseidonAnnotation, H>,
+}
+
+impl<T, H> PoseidonTree<T, H>
+where
+    T: Content<H>,
+    for<'any> &'any T: Into<StorageScalar>,
+    H: ByteHash,
+{
+    /// Constructs a new empty PoseidonTree
+    pub fn new(depth: usize) -> Self {
+        PoseidonTree {
+            branch_depth: depth,
+            inner: Default::default(),
+        }
+    }
+
+    /// Returns the scalar root-hash of the poseidon tree
+    ///
+    /// This includes padding the value to the correct branch length equivalent
+    pub fn root(&self) -> io::Result<Scalar> {
+        if let Some(ann) = self.inner.annotation() {
+            let borrow: &StorageScalar = ann.borrow();
+            let scalar: Scalar = borrow.clone().into();
+
+            // FIXME, depth could be inferred from the cardinality
+            if let Some(branch) = self.get(0)? {
+                let depth = branch.levels().len();
+                Ok(extend_scalar(scalar, self.branch_depth - depth))
+            } else {
+                unreachable!("Annotation in empty tree")
+            }
+        } else {
+            // empty case, use an empty level for hashing
+            let leaves = [Scalar::zero(); ARITY + 1];
+            let level = PoseidonLevel { leaves, offset: 0 };
+            let root = merkle_level_hash_without_bitflags(&level);
+            Ok(extend_scalar(root, self.branch_depth))
+        }
+    }
+
+    /// Returns a poseidon branch pointing at the specific index
+    ///
+    /// This includes padding the value to the correct branch length equivalent
+    pub fn poseidon_branch(
+        &self,
+        idx: u64,
```

Comment:
Shouldn't it ask for the height? So it can know how much should it pad?

Reply:
Nevermind, you have it on the PoseidonTree stored. My bad

Reply:
It pads based on the self.branch_depth, no reason to pass it here.